### PR TITLE
Modular routing

### PR DESCRIPTION
**Pull request by jroweboy**
Finished modular routing and merged the reverse routing Robert wrote into it as well. There are a few TODOs in reverse routing left, but its is "working" condition minimally. From here on out, I'll mostly be providing advice on rust to my group members that don't know much rust, so commits from me will slow down until the end of the semester.

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -7,3 +7,6 @@
 [submodule "rust-mustache"]
 	path = rust-mustache
 	url = https://github.com/jroweboy/rust-mustache.git
+[submodule "rust-postgres"]
```
Comment (LeviSchuck) [inline .gitmodules:10]: This really should have been in a separate pull request. It is not related to routing.

File: example/hello_world/index.rs [hunk 2]
```diff
@@ -1,71 +1,65 @@
 extern crate oxidize;
 extern crate collections;
 
-use oxidize::{Oxidize, Request, Response, Route};
+use oxidize::{Oxidize, Request, ResponseWriter, Config, MediaType};
```
Comment (LeviSchuck) [inline example/hello_world/index.rs:4]: This might be ignorance, but why not just do `use oxidize`?

File: example/hello_world/index.rs [hunk 3]
```diff
@@ -1,71 +1,65 @@
...
 
-
 fn main() {
-    // TODO: rework the params to be more sensible :p 
-    // See oxidize.rs Oxidize struct for more info
-    let server = Oxidize::new(8000, "localhost", routes);
+    // TODO capture groups currently clash :S I need to rework that a bit
+    let routes: ~[RegexRoute] = ~[
+        Regex(Route{ method: "GET", name: "index", path: "^/$", fptr: index}),
+        Regex(Route{ method: "GET", name: "test_mustache", path: "^/test/?$", fptr: test_mustache}),
```
Comment (LeviSchuck) [inline example/hello_world/index.rs:51]: So, these strings are not owned `~"taco"`, why?

I feel like we could have a `regexIndex` constant instead of making users type `"^/$"`, but couldn't it work in `Simple()` as well with just `"/"`?

File: rust-mustache [hunk 4]
```diff
@@ -1 +1 @@
-Subproject commit 7414d9917fc79fb5bbac24c45d521c513bd81db0
+Subproject commit 1955e106d8ea83c4132a271b6875fe467f8796a6
```
Comment (LeviSchuck) [inline rust-mustache:1]: Likely not related to this pull

File: rust-postgres [hunk 5]
```diff
@@ -0,0 +1 @@
+Subproject commit 7c94c206718cff06b5ddefe6c02314e3478ebbdf
```
Comment (LeviSchuck) [inline rust-postgres:1]: Likely not related to this pull.

File: src/renderer.rs [hunk 6]
```diff
@@ -1,59 +1,33 @@
...
-    // println!("Render for this file: {}", path.display());
-    // let contents = File::open(&path).read_to_end();
-
-    // match from_utf8(contents.unwrap()) {
-    //     Some(str) => str,
-    //     None => fail!("File could not be rendered")
-    // }
     let path = Path::new("templates/"+file_name);
-    println!("Render for this file: {}", path.display());
+    debug!("Render for this file: {}", path.display());
```
Comment (LeviSchuck) [inline src/renderer.rs:13]: I see you took note of my questioning

File: src/renderer.rs [hunk 7]
```diff
@@ -1,59 +1,33 @@
...
-    // TODO: add the request to the context so that they can use things like session vars
+    // TODO: add the request to the context so that they can use things like session vars?
     from_utf8_owned(file_contents).expect("File could not be parsed as UTF8")
 }
 
-#[allow(unused_variable)]
-// , T: Encodable<Encoder> , context: Option<T>
-pub fn mustache_render<'a>(request: &'a mut Request, file_name: &'a str, 
+// ditto above request: &'a mut Request, 
+pub fn mustache_render<'a>(file_name: &'a str,
```
Comment (LeviSchuck) [inline src/renderer.rs:20]: I do not have enough rust experience, what does `'a` mean?

Comment (jroweboy) [inline src/renderer.rs:20]: Its a lifetime specifier. It means that the life time of the &'a are guaranteed to at least be as long as the function call.

File: src/route/mod.rs [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+    fn copy(&self) -> ~Router;
+}
+
+impl Clone for ~Router {
+    fn clone(&self) -> ~Router {
+        self.copy()
+    }
+}
+
+pub mod regexrouter;
```
Comment (LeviSchuck) [inline src/route/mod.rs:18]: Some people complain if you don't have an empty line at the end of the source code

File: src/route/mod.rs [hunk 9]
```diff
@@ -0,0 +1,18 @@
+use request::Request;
+use http::server::ResponseWriter;
+use collections::hashmap::HashMap;
+
+pub trait Router {
+    // TODO: a router should have a way to call the middleware stack before and after the view
```
Comment (LeviSchuck) [inline src/route/mod.rs:6]: Why?

Comment (jroweboy) [inline src/route/mod.rs:6]: it can either go there or it can be called from oxidize. It doesn't matter to me, but while I was having the thought I wrote a comment about it

File: src/route/regexrouter.rs [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+                    Some(i) => {
+                        self.routes.read( |x: &~[RegexRoute]| { 
+                            match x[i] {
+                                Simple(s) => {(s.fptr)(request, response); Some("success")},
+                                Regex(s) => {(s.fptr)(request, response); Some("success")},
+                                Static(_) => {debug!("TODO: impl StaticServe"); None},
+                            }
+                        }
+                    )}
+                    None => None
```
Comment (LeviSchuck) [inline src/route/regexrouter.rs:47]: See, this is why applicatives in Haskell exist.

File: rust-postgres (hunk 5 above)
Comment (jroweboy) [inline rust-postgres:1]: While I would typically agree that I shouldn't put unrelated things into the repository like that, in this instance and as I kinda hinted at in my pull request comment, the point of this commit is to update the repo with all my changes (regardless of what they were related to) and focus on getting others up to speed. In a normal case  I agree that I should move them to another pull request though

File: src/route/regexrouter.rs [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+            }
+        };
+    }
+
+    #[allow(unused_variable)]
+    fn reverse<'a>(&'a self, name: &str, vars: Option<HashMap<~str,~str>>) -> Option<&'a ~str> {
+        // TODO use the vars to replace regex things 
+        // TODO remove all the ugly regexy stuff to make a valid URL
+        self.reverse_routes.get().find_equiv(&name).and_then(
+            |path: &'a ~str| { Some(path) }
```
Comment (LeviSchuck) [inline src/route/regexrouter.rs:58]: Such symbols, much non-alpha-numeric, wow

File: src/route/regexrouter.rs [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+        }
+
+        regex.push_str("|");
+        // TODO add the method to the regex
+        //regex.push_str(route.method.to_owned());
+        // debug!("About to unwrap {}", route_regex);
+        regex.push_str(route_regex.unwrap().to_owned());
+        regex.push_str("(*MARK:");
+        regex.push_str(i.to_str());
+        regex.push_str(")");
```
Comment (LeviSchuck) [inline src/route/regexrouter.rs:112]: is push_str an O(1) operation?

File: src/route/regexrouter.rs [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+}
+
+pub struct Serve {
+    method: &'static str,
+    path: &'static str,
+    directory: &'static str, 
+}
+
+
+// TODO make a way to import routes from another app
```
Comment (LeviSchuck) [inline src/route/regexrouter.rs:168]: Use Colons after TODO, IDEA, etc.

File: src/route/regexrouter.rs [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+impl Serve { 
+    #[allow(dead_code)]
+    fn get_star_regex(&self) -> ~str {
+        let mut regex = ~"^";
+        regex.push_str(self.path.replace("*", ".+"));
+        regex.push_char('?');
+        regex
+    }
+
+    #[allow(dead_code)]
```
Comment (LeviSchuck) [inline src/route/regexrouter.rs:191]: What prompted this?

Comment (jroweboy) [inline src/route/regexrouter.rs:191]: I'm not sure I understand what you mean by that.

File: src/route/regexrouter.rs [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+    #[allow(unused_must_use)]
+    fn load_file(path: &str, r: &mut ResponseWriter) {
+        debug!("Render for this file: {}", path);
+        let file_path = Path::new(path);
+        match File::open(&file_path).read_to_end() {
+            Ok(s) => {
+                r.status = status::Ok;
+                r.headers.content_type = 
+                    Some(MediaType {type_: ~"image", subtype: ~"jpg", parameters: ~[]});
```
Comment (LeviSchuck) [inline src/route/regexrouter.rs:200]: So, are you hard coding all files to return as jpegs?

Comment (jroweboy) [inline src/route/regexrouter.rs:200]: Static file serve is not implemented yet (and even if you try to use it it will just print debug messages), but I wrote the start to a few of the key functions involved and yes, as an example I return all files as jpgs :smile:

File: src/route/mod.rs (hunk 9 above)
Comment (LeviSchuck) [inline src/route/mod.rs:6]: No, Why "should [it] have a way to call the middleware stack before and after"?

File: example/hello_world/index.rs (hunk 2 above)
Comment (jroweboy) [inline example/hello_world/index.rs:4]: its just one of those things where I don't like having to write oxidize::Request oxidize::ResponseWriter everywhere :wink:

File: example/hello_world/index.rs (hunk 3 above)
Comment (jroweboy) [inline example/hello_world/index.rs:51]: Yeah, I can always append prepend ^ $ to the Regex(), I just currently don't. I'll probably add that and update this commit then.

File: example/hello_world/index.rs (hunk 2 above)
Comment (LeviSchuck) [inline example/hello_world/index.rs:4]: Does rust not 'import' what the `oxidize` module exports when you just leave it at `use oxidize`?

File: src/route/regexrouter.rs (hunk 14 above)
Comment (LeviSchuck) [inline src/route/regexrouter.rs:191]: Why did you add a pragma to allow for dead code?

File: src/route/mod.rs (hunk 9 above)
Comment (jroweboy) [inline src/route/mod.rs:6]: Its just a comment :cold_sweat: (pls no hurt me) I'm not saying that we will do it that way haha (glad you aren't reading some of my old TODO comments then) I'm always open to changing the way things work.

File: src/route/regexrouter.rs (hunk 12 above)
Comment (jroweboy) [inline src/route/regexrouter.rs:112]: It looks like its O(expand underlying array and then memcpy) so most likely O(n) where n is source

```
#[inline]
pub fn push_bytes(dst: &mut ~[u8], src: &[u8]) {
    let old_len = dst.len();
    dst.reserve_additional(src.len());
    unsafe {
        ptr::copy_memory(dst.as_mut_ptr().offset(old_len as int), src.as_ptr(), src.len());
        dst.set_len(old_len + src.len());
    }
}
```

File: src/route/regexrouter.rs (hunk 14 above)
Comment (jroweboy) [inline src/route/regexrouter.rs:191]: Cause it is just dead code since I don't call it from anywhere :smiley_cat: all the `#[allow(dead_code)]` does is suppresses warnings on it. I put that code in there as an example of how we will need to convert the "/static/*" into a regex, but since its not used I wanted to avoid warnings that I know about :rabbit2:

File: rust-postgres (hunk 5 above)
Comment (robhobbes) [inline rust-postgres:1]: Yeah, I think it's pretty normal to not have perfectly broken down commits early on in the project. We're just trying to make headway and @jroweboy had so many sweeping changes that it would have been hard for him to micro all the different pull requests. So, I don't think it is a huge deal that he accidentally included a few slightly unrelated things this time.

File: src/route/mod.rs (hunk 8 above)
Comment (robhobbes) [inline src/route/mod.rs:18]: At some point (probably, when we figure out our code style) we should do a single "code cleanup" pull request. Until then, let's not worry about it.

File: example/hello_world/index.rs (hunk 2 above)
Comment (jroweboy) [inline example/hello_world/index.rs:4]: there are blob imports `use oxidize::*` but they are highly frowned upon just like in python. `use oxidize` does import all the `oxidize` exports in the sense that you can then call `oxidize::Whatever`

File: src/route/mod.rs (hunk 8 above)
Comment (jroweboy) [inline src/route/mod.rs:18]: We should have that added to the Gantt chart. I think that would be a cool thing to do. We are about to have a documentation and testing push here in a bit as well.

File: src/route/mod.rs (hunk 9 above)
Comment (LeviSchuck) [inline src/route/mod.rs:6]: Well, I'm concerned about the architecture here, and it does not seem intuitive to me to have the router need access to the middleware stack, it should be the other way around.

File: src/route/regexrouter.rs (hunk 12 above)
Comment (LeviSchuck) [inline src/route/regexrouter.rs:112]: Remind me, is this merely on constructing the routes and thus only impacts (milliseconds of) startup time , or is it executed upon every request?

File: src/route/mod.rs (hunk 9 above)
Comment (jroweboy) [inline src/route/mod.rs:6]: As I said, this is not a WE MUST DO. This is a suggestion. Please consider it such. I will remove this comment since its caused so much grief already.

File: src/route/regexrouter.rs (hunk 12 above)
Comment (jroweboy) [inline src/route/regexrouter.rs:112]: Once at startup.

File: example/hello_world/index.rs (hunk 2 above)
Comment (robhobbes) [inline example/hello_world/index.rs:4]: Yeah, if you explicitly do `use oxidize::{Oxidize, Request, Route}`, then you can use `Route` without doing `oxidize::Route`. If you just do `use oxidize` then to use `Route` you have to say `oxidize::Route`.

File: src/route/mod.rs (hunk 8 above)
Comment (robhobbes) [inline src/route/mod.rs:18]: @trevordixon You should look at this thread...

File: src/route/regexrouter.rs (hunk 13 above)
Comment (robhobbes) [inline src/route/regexrouter.rs:168]: I agree that that is probably a good practice.

File: src/route/regexrouter.rs (hunk 12 above)
Comment (LeviSchuck) [inline src/route/regexrouter.rs:112]: Then my concern has been resolved.

File: example/hello_world/index.rs [hunk 16]
```diff
@@ -1,71 +1,65 @@
...
+    ];
+
+    let router = ~RegexRouter::new(routes);
+    let conf = Config {
+        debug: true,
+        bind_addr: from_str::<SocketAddr>("127.0.0.1:8001").unwrap(),
+    };
+
+    let server = Oxidize::new(conf, router as ~Router);
     server.serve();
```
Comment (robhobbes) [inline example/hello_world/index.rs:64]: Would it be possible (or desirable) to make the constructor for Oxidize call `serve` on itself so that the user doesn't have to explicitly call `server.serve()`

Comment (jroweboy) [inline example/hello_world/index.rs:64]: I don't know if that's desirable or not. It is possible though since `new` isn't a constructor in the strictest sense. Its just a static function on Oxidize. Most other frameworks seem to do it this way though

**Comment by jroweboy** [comment]
I'm recreating this branch and committing the changes that I have made to bring it in line with the comments presented here.
